### DISCUSSION TITLE
Add a JSON.parse function which can receive re-usable objects from an object pool

### DISCUSSION THREAD

**Discussion by jdgriffith** [oven-sh/bun#7570] [Ideas]
In certain high ingestion scenarios from a websocket, it might be advantageous to use an object pool of pre-initialized objects which can be used for JSON parsing where a new object does not need to be created. This could help in several ways as it could help tell the JSON parser the shape of the data and also no new memory needs to be allocated for the object. Once an object is finished being utilized by processing, it could be given back to the object pool. I'm not asking for Bun to create an API for an object pool (but I'm not opposed!) but I am asking for more logic on the JSON processing side - the ability to pass an object along with the JSON string.

```
JSON.parse(jsonString, objectToBePopulated)
```

or

```
JSON.parse(jsonString, objectPool)
```

I'm curious if my thought about this helping with performance would hold true in this scenario but it's an interesting thought. I wonder what others think of this proposal.

**Comment by guest271314** [comment]
You can already do that by spreading data into an object then using `JSON.stringify()` on the object before/at passing to `JSON.parse()`. Note, the second parameter to `JSON.parse()` is a "reviver" function and with `JSON.stringify()` is a "replacer" function, both of which are recursive where you can mutate the result.

**Reply by guest271314** [reply]
> This could help in several ways as it could help tell the JSON parser the shape of the data and also no new memory needs to be allocated for the object. Once an object is finished being utilized by processing, it could be given back to the object pool.

You can write the data to a resizable `ArrayBuffer`, get and process the data using `DataView` then resize the `ArrayBuffer` to `0`.

Note, there is also `WebSocketStream` where no data has to be stored at all.

```
var wss = new WebSocketStream('wss://path/to/server');
var {
  readable,
  writable
} = await wss.opened;
readable.pipeThrough(new TextDecoderStream()).pipeTo(
  new WritableStream({
    write(value) {
      // Do stuff with streamed data
    }
  })
);
```

**Reply by jdgriffith** [reply]
Thanks for your response! 

I'm familiar with the reviver function but it still doesn't address the top level concern (from my understanding). The reviver function allows you to serialize or transform individual key values but does not give you control over the object created or used. 

Streams and buffers are great, but you still have to serialize that data at some point using JSON.parse or some other JSON parser and convert into objects which are usable within the runtime.

**Reply by guest271314** [reply]
Depends on the use case and API's being used if `JSON` is necessary. With audio and video we can pass `Uint8Array` `buffer` to a different `TypedArray`. If we are streaming over IPC, yes, then we can use `JSON.parse()` to get `[0, 255,...]` from JSON `"[0, 255]"`.

With a resizable `ArrayBuffer` you ca use then re-sed the same `ArrayBuffer`. With a `SharedArrayBuffer` you can store your JSON data there, retrieve data, overwrite data in the same buffer. 

I have streamed JSON over IPC without any issues. I've also written to the same resizable `ArrayBuffer`, and streamed using `Uint8Array`s. 

I guess I don't know what you mean by "control over the object created or used" or "object pool", in relation to `JSON.parse()`.

**Reply by jdgriffith** [reply]
The point of this post is a very specific use case. It's about receiving `strings` from a websocket (or other protocol from a third party) and needing to convert those strings into JS objects. I don't have these "objects" yet in memory in any form. It's interesting you mention audio and video processing but it's not relevant to the very specific use case I mentioned to start the discussion.

The use case is `websocket` -> `strings` -> `deserialization` -> `JS Objects`. The deserialization is the most important part and usually involves `JSON.parse`. `JSON.parse` _creates_ objects within it's logic, which is the heart of the matter, and it doesn't expose an interface to control this. Instead of `JSON.parse` creating objects, I would like to pass an existing object to `JSON.parse` or some other deserialization library.

If by chance I missed something in what you were trying to say, please post a complete code snippet, including the deserialization process. I'm all for learning new performance tricks.

**Reply by guest271314** [reply]
> and it doesn't expose an interface to control this.

It does though. The `reviver` function. 

> [reviver](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#reviver) Optional
>
> If a function, this prescribes how each value originally produced by parsing is transformed before being returned. Non-callable values are ignored. The function is called with the following arguments:
> 
>> [key](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#key)
> The key associated with the value.
> 
>> [value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#value)
> The value produced by parsing.

You don't have to use the returned vJalue of `JSON.parse()`. You can just use `parse()` to parse and ignore any return value, by just referencing the object you already have, using default parameters to `reviver`, or using `bind()`, with `reviver`. E.g.

```
var o = {b:1};
function reviver(key, value, {source}) {
  // Do what you need to do with key, value, or source
  // spread ingto to o, ignoring return value of parse
 console.log({key, value, source, o});
}

JSON.parse(`"a"`, reviver);
```

**Reply by guest271314** [reply]
>  I'm all for learning new performance tricks.

I'm not sure how performance is involved?

**Reply by jdgriffith** [reply]
I think you might be still missing my original point. `JSON.parse` is _creating_ an object under the hood, even when using the `reviver` function. This can be a performance hit in a high ingestion rate scenario (thousands to a million messages per second). I appreciate you jumping in with your thoughts, but we have to agree on this to get to valid solutions.

**Reply by guest271314** [reply]
> This can be a performance hit in a high ingestion rate scenario (thousands to a million messages per second).


No, it's not. 

I stream real-time PCM at 44100 samples per second. 

Youv will will notice if something is happening with your code by gaps and glitches in the real-time stream.

I don't have  gaps and  glitches in the stream that originates as JSON.

Performance issues doesn't happen just because `JSON.parse()` is involved. 

I would need to see evidence and reproduce where `JSON.parse()` alone is causing performace issues.

Some real-world code that works:

This is sending JSON to the browser from QuickJS https://github.com/guest271314/captureSystemAudio/blob/10908bd48f08018a338aa046a26429189bbcdc7b/native_messaging/capture_system_audio/capture_system_audio.js#L21C1-L34C2

```
function main() {
  const message = getMessage();
  const size = 1764;
  let data = new Uint8Array(size);
  const pipe = std.popen(
    JSON.parse(String.fromCharCode(...message)),
    'r'
  );
  while (pipe.read(data.buffer, 0, data.length)) {
    sendMessage(`[${data}]`);
    pipe.flush();
    std.gc();
  }
}
```

this is parsing that JSON in real-time https://github.com/guest271314/captureSystemAudio/blob/10908bd48f08018a338aa046a26429189bbcdc7b/native_messaging/capture_system_audio/transferableStream.js#L14-L25

```
  async function handleMessage(value, port) {
    if (!Array.isArray(value)) {
      value = JSON.parse(value);
    }
    try {
      await writer.ready;
      await writer.write(new Uint8Array(value));
    } catch (e) {
      console.error(e.message);
    }
    return true;
  }
```

**Reply by guest271314** [reply]
You can read Chromium source code for how the Native Messaging protocol (over IPC, which is string-based) is implemented. 

https://developer.chrome.com/docs/extensions/mv3/nativeMessaging/#native-messaging-host-protocol

> Chrome starts each native messaging host in a separate process and communicates with it using standard input (stdin) and standard output (stdout). The same format is used to send messages in both directions; each message is serialized using JSON, UTF-8 encoded and is preceded with 32-bit message length in native byte order. The maximum size of a single message from the native messaging host is 1 MB, mainly to protect Chrome from misbehaving native applications. The maximum size of the message sent to the native messaging host is 4 GB.

**Reply by jdgriffith** [reply]
I do think about performance at the lowest levels. I have to because I'm operating/creating a high-frequency trading system (HFT). Your 44k of messages is not inconsequential, but I'm looking at scales that could reach **1 million** msg/second. You can see in the benchmarks below I'm looking closely at how long at takes to parse JSON in different ways, create an object, populate an object, etc. 

However, I don't want this to devolve into a debate/argument about performance, which can get skewed if it doesn't remain on topic. This is a post mostly to the creators of bun about the possibility of creating a larger interface for JSON. And respectfully, I don't want to argue about this. 

![Screenshot 2023-12-12 at 8 52 11 AM](https://github.com/oven-sh/bun/assets/872496/5dd92707-3abc-4a43-8c5e-fb869ee6b803)

**Reply by guest271314** [reply]
I've found that using `[0, 255,...]` is the simplest format, which can be used in C, C++, et al. Then we can construct a `Uint8Array` and use `TextDecoder()` if necessary.

I have not encountered any issues streaming real-time audio, and writing the raw PCM to an MP3 in real-time. It took a lot of testing to achieve a media stream that can be shared with peers as a `MediaStreamTrack`, and played back at the same time.

I'm wondering what the user observes given your tests? And what type of parser you are proposing, in source code? And if what you propose will be different from `JSON.parse()`.

I'll fall back and observe what you come up with, and other feedback. Then test. Good luck!

**Reply by jdgriffith** [reply]
Thanks! Maybe something interesting will come from this 😄

**Reply by guest271314** [reply]
I really don't think JSON is a problem, performance wise. Though if you do you're gonna have to re-write a JSON parser. Honestly I would just use `WebSocketStream` or `fetch()` and pipe through a `TextDecoderStream` or use a `TransformStream` directly where if you are in control of the data you can pass the data to `Response.json()` in the transform method, then you'll be streaming data and handling `JSON` in, something like

```
var {readable, writable} = new TransformStream();
readable /* (await fetch("/hv/day/trading/app")).body */ .pipeThrough(new TransformStream({
  async transform(v, c) {
    c.enqueue(await new Response(v).json())
  }
})).pipeTo(
  new WritableStream({
    write(v) {
      console.log(v)
    }
  })
);

var writer = writable.getWriter();
writer.write(`{"a":123}`)
```

Again, good luck. I'm curious to see how you write out the specification.

**Reply by guest271314** [reply]
Here's QuickJS's Binary JSON https://bellard.org/quickjs/quickjs.html#Binary-JSON https://github.com/bellard/quickjs/blob/master/tests/test_bjson.js, https://github.com/bellard/quickjs/blob/master/tests/bjson.c.

**Reply by jdgriffith** [reply]
I'm not sure how to communicate this to you - it's not about streams or the surrounding interfaces for transmitting or piping the data. Even your example above probably uses `JSON.parse` under the hood in `new Response(v).json()`, if that returns an object. I use streams all the time for large amounts of data and they are wonderful for that - super efficient and fast. 

My post and performance concerns are _only_ about `JSON.parse` and that's it. Plenty of other projects have explored the concept of super efficient JSON parsers, including using SIMD, and I'm not the first to explore more efficiency as it relates to JSON processing.

**Reply by guest271314** [reply]
What do you suggest? How would you specify the algorithm?

I'm curious why `JSON.parse` is important for your application? 

I've used various means to parse JSON, including `RegExp`.

No matter what you do you are going to have to parse everything between opening and closing `"`. I just don't see how you can come up with something faster than what exists, given the mandatory condition above.

**Reply by guest271314** [reply]
For some context I day traded 20 years ago so I know those seconds mean something. I just don't see how you can reliably speed up JSON parsing when the parser has to handle nested objects and arrays.

**Reply by guest271314** [reply]
Maybe farm out JSON parsing to WebAssembly or something?

If you are proposing a new feature to the language I would write out the steps of the algorithm, perhaps including spreading into an existing object and branching then breaking the flow when specific properties or values are found during parsing, which `JSON.parse()` implements, e.g., https://stackoverflow.com/a/39263212

```
var [title, innerData] = JSON.parse(JSON.stringify([data.title, data.innerData], ["title"]));
```

otherwise you are going to have to parse the entire string, it's mandatory.

**Reply by guest271314** [reply]
A `for` loop is faster starting from the end than from the beginning, so in theory and practice you can use a `for` loop on the JSON string and `break` when specific property names of objects are matched. 

Using arrays is far simpler, if you have control over the input and the data does not include `{"property":"value"}`.
